_practicalities/intro.md

```diff
@@ -0,0 +1,355 @@
+install . `  and `pip install -e .` for regular and developer install.
+
+Check in scripts, and documentation that the correct installation command is
+used. 
+
+# Recommended Mitigations
+
+These are not mandatory but should make the transition seamless by warning your
+user early enough _and_ providing useful error messages.
+
+## Runtime warning on master
+
+Add a warning at _runtime_ early on master (before switching to Python 3
+  only)
+
+```
+import warnings
+import sys
+if sys.version_info < (3,):
+    warnings.warn('You are using master of `Frobulator` with Python 2. '
+                  'Frobulator will soon be Python 3 only. '
+                  'See this issue to know more.',
+                  UserWarning)
+```
+
+Your Python 2 user have a chance to upgrade, or get off master, (for example on
+the LTS branch).
+
+## Fail early at import time
+
+Add an error early at import at runtime with a clear error message, leave the
+early import compatible Python 2 for users to not be welcomed with a useless
+`SyntaxError`. You are _allowed_ to use multi-line strings in error messages.
+
+Error at import time _will_ happen on system with old version of pip and
+setuptools. Keep in mind that saying the package is Python 3 only is not a lot
+more helpful than a Syntax error. The most reasonable reason would be out of
+data pip and setuptools:
+
+
+```
+import sys
+
+if sys.version_info < (3,):
+    raise ImportError(
+    """You are running Frobulator 6.0 on Python 2
+
+Unfortunately Frobulator 6.0 and above are not compatible with Python 2
+anymore, and you still ended up with this version installed on your system.
+That's a bummer. Sorry about that. It should not have happen. Make sure you
+have pip >= 9.0 to avoid this kind of issues, as well as setuptools >= 24.2:
+
+ $ pip install pip setuptools --upgrade
+
+You have various other choices
+
+- install an older version of Frobulator:
+
+ $ pip install 'frobulator<6.0'
+
+- Upgrade your system to use Python 3.
+
+It would be great if you can figure out how this version ended up being
+installed, and try to check how to prevent that for future users.
+
+See the following url for more up to date informations:
+
+https://i.am.an/url
+
+""")
+
+```
+
+## Watch out for beta releases
+
+
+Make sure your version number match pep 440 or you will get surprises during
+beta in particular as the `sdist` and `wheel` will appear as being different
+versions, in particular sdist (during beta/rc/post) can appear with a greater
+version number than wheels. Pip thus try to install the sdist instead of the
+wheel, which have more chance of failing, in particular with pre 24.2 versions
+of setuptools.
+
+The regular expression to check for validity of pep440 can be find below:
+
+  `^([1-9]\\d*!)?(0|[1-9]\\d*)(\\.(0|[1-9]\\d*))*((a|b|rc)(0|[1-9]\\d*))?(\\.post(0|[1-9]\\d*))?(\\.dev(0|[1-9]\\d*))?`
+
+## Depend on setuptools
+
+You can mark your library as dependent on setuptools greater than 24.3 starting
```

Comment:
I'm reluctant to recommend this because most libraries do not actually depend on setuptools, and it's annoying for packaging systems that install packages another way. With wheels, the possibility of having virtualenvs without setuptools installed is not that far off. Can we at least put in a caveat?